### DISCUSSION TITLE
How to use the `check` functionality outside of the `IOChecker` test class

### DISCUSSION THREAD

**Discussion by yatesco** [tpolecat/doobie#1621] [Q&A] [answered]
Hi, I understand that `doobie.munit.IOChecker` exposes a `check` function which can be used to statically type check the query. However, `IOChecker` also requires a static `transactor` which assumes a single `transactor` for all the tests in that test suite:
```scala
class DoobieDaoTestSuite extends FunSuite with doobie.munit.IOChecker {

  override val colors = doobie.util.Colors.None // just for docs

  val transactor = Transactor.fromDriverManager[IO](
    "com.microsoft.sqlserver.jdbc.SQLServerDriver", "jdbc:sqlserver://localhost;integratedSecurity=false;TrustServerCertificate=true;databaseName=XXXX", "sa", "XXXX"
  )

  val dao = new DoobieDao()

  test("trivial")    { check(dao.loadAll())        }

}
```

This model is incompatible with the https://github.com/testcontainers/testcontainers-scala model of exposing the container in a closure, per test:
```scala
class DoobieDaoTestContainersTestSuite extends FunSuite with TestContainerForEach {

//  override val containerDef: MSSQLServerContainer.Def = MSSQLServerContainer.Def(urlParams = Map("databaseName" -> "pride-and-joy-next"))
  override val containerDef: MSSQLServerContainer.Def = MSSQLServerContainer.Def()

  val dao = new DoobieDao()

  test("trivial") {
    withContainers(container => {
      val transactor = Transactor.fromDriverManager[IO](
        container.driverClassName, container.jdbcUrl, container.username, container.password
      )
      val y = transactor.yolo
      import y._

      val query = dao.loadAll();
      query.check.unsafeRunSync()
    })
  }

}
```

Unfortunately, calling `query.check` doesn't fail, even if it doesn't check.

Can you please give me some guidance on how to proceed, so I can use the `check` functionality inside tests that don't extend `IOChecker`.

**Comment by jatcwang** [comment, marked as answer]
Hi @yatesco I would recommend looking at how `check` is [implemented](https://github.com/tpolecat/doobie/blob/f800ec1a89120c16ca6656075b592dbf28642fd7/modules/munit/src/main/scala/doobie/munit/analysisspec.scala#L51) in Checker/IOChecker trait. 

The idea would be to implement your own `check` which takes an implicit  `transactor: Transactor[M]` argument and use that in `val report = U.unsafeRunSync(analyze(args).transact(transactor))`.

**Reply by yatesco** [reply]
Thanks!
